FutureFlattenSourceSpec: handle downstream cancelling before the underlying Future completes

```
[ERROR] [09/19/2017 04:31:10.939] [FutureFlattenSourceSpec-akka.test.stream-dispatcher-9] [akka://FutureFlattenSourceSpec/user/StreamSupervisor-279/flow-8-0-terminationWatcher] null
java.lang.NullPointerException
	at akka.event.LogSource$$anon$6.genString(Logging.scala:283)
	at akka.event.LogSource$$anon$6.genString(Logging.scala:282)
	at akka.event.LogSource$.apply(Logging.scala:306)
	at akka.event.Logging$.apply(Logging.scala:557)
	at akka.stream.impl.fusing.GraphInterpreterShell.log$lzycompute(ActorGraphInterpreter.scala:448)
	at akka.stream.impl.fusing.GraphInterpreterShell.log(ActorGraphInterpreter.scala:448)
	at akka.stream.impl.fusing.GraphInterpreterShell.interpreter$lzycompute(ActorGraphInterpreter.scala:483)
	at akka.stream.impl.fusing.GraphInterpreterShell.interpreter(ActorGraphInterpreter.scala:483)
	at akka.stream.impl.fusing.GraphInterpreterShell.toString(ActorGraphInterpreter.scala:626)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1$$anonfun$applyOrElse$2.apply(ActorGraphInterpreter.scala:763)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1$$anonfun$applyOrElse$2.apply(ActorGraphInterpreter.scala:761)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:761)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:517)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:650)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
	at akka.actor.ActorCell.invoke(ActorCell.scala:496)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

--- Stream actors debug dump ---
Stream is completed. No debug information is available
[info] - must handle downstream cancelling before the underlying Future completes *** FAILED *** (7 seconds, 50 milliseconds)
[info]   java.lang.AssertionError: assertion failed: expected no StreamSupervisor children, but got [Actor[akka://FutureFlattenSourceSpec/user/StreamSupervisor-279/flow-8-0-terminationWatcher#1909747338]]
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(Utils.scala:31)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply(Utils.scala:28)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply(Utils.scala:28)
[info]   at akka.testkit.TestKitBase$class.poll$2(TestKit.scala:302)
[info]   at akka.testkit.TestKitBase$class.awaitAssert(TestKit.scala:313)
[info]   at akka.testkit.TestKit.awaitAssert(TestKit.scala:835)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply$mcV$sp(Utils.scala:28)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply(Utils.scala:26)
[info]   at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply(Utils.scala:26)
[info]   at akka.testkit.TestKitBase$class.within(TestKit.scala:344)
[info]   at akka.testkit.TestKit.within(TestKit.scala:835)
[info]   at akka.testkit.TestKitBase$class.within(TestKit.scala:358)
[info]   at akka.testkit.TestKit.within(TestKit.scala:835)
[info]   at akka.stream.testkit.Utils$.assertAllStagesStopped(Utils.scala:26)
[info]   at akka.stream.scaladsl.FutureFlattenSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$5.apply(FutureFlattenSourceSpec.scala:81)
[info]   at akka.stream.scaladsl.FutureFlattenSourceSpec$$anonfun$1$$anonfun$apply$mcV$sp$5.apply(FutureFlattenSourceSpec.scala:81)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[info]   at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
[info]   at akka.stream.testkit.StreamSpec.withFixture(StreamSpec.scala:26)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:1075)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:1088)
[info]   at akka.testkit.AkkaSpec.runTest(AkkaSpec.scala:59)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.WordSpecLike$class.runTests(WordSpecLike.scala:1147)
[info]   at akka.testkit.AkkaSpec.runTests(AkkaSpec.scala:59)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1147)
[info]   at akka.testkit.AkkaSpec.org$scalatest$WordSpecLike$$super$run(AkkaSpec.scala:59)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.WordSpecLike$class.run(WordSpecLike.scala:1192)
[info]   at akka.testkit.AkkaSpec.org$scalatest$BeforeAndAfterAll$$super$run(AkkaSpec.scala:59)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
[info]   at akka.testkit.AkkaSpec.run(AkkaSpec.scala:59)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:480)
[info]   at sbt.TestRunner.runTest$1(TestFramework.scala:76)
[info]   at sbt.TestRunner.run(TestFramework.scala:85)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFunction.apply(TestFramework.scala:207)
[info]   at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$4.work(System.scala:63)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[info]   at sbt.Execute.work(Execute.scala:237)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
[info]   at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:745)
```

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
@@ -92,6 +92,9 @@ class FutureFlattenSourceSpec extends StreamSpec {
       probe.request(1)
       probe.cancel()
 
+      // try to avoid a race between probe cancel and completing the promise
+      Thread.sleep(100)
+
       // even though canceled the underlying matval should arrive
       sourcePromise.success(underlying)
       val failure = sourceMatVal.failed.futureValue

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
@@ -99,7 +99,7 @@ class FutureFlattenSourceSpec extends StreamSpec {
       sourcePromise.success(underlying)
       val failure = sourceMatVal.failed.futureValue
       failure shouldBe a[StreamDetachedException]
-      failure.getMessage should ===("Stream finished before future source completed")
+      failure.getMessage should ===("Stream cancelled before Source Future completed")
     }
 
     "fail if the underlying Future is failed" in assertAllStagesStopped {

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
@@ -94,7 +94,9 @@ class FutureFlattenSourceSpec extends StreamSpec {
 
       // even though canceled the underlying matval should arrive
       sourcePromise.success(underlying)
-      sourceMatVal.failed.futureValue.getMessage should ===("Stream finished before future source completed")
+      val failure = sourceMatVal.failed.futureValue
+      failure shouldBe a[StreamDetachedException]
+      failure.getMessage should ===("Stream finished before future source completed")
     }
 
     "fail if the underlying Future is failed" in assertAllStagesStopped {

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala
@@ -81,18 +81,20 @@ class FutureFlattenSourceSpec extends StreamSpec {
     "handle downstream cancelling before the underlying Future completes" in assertAllStagesStopped {
       val sourcePromise = Promise[Source[Int, String]]()
 
-      val (sourceMatVal, termination) =
+      val probe = TestSubscriber.probe[Int]()
+      val sourceMatVal =
         Source.fromFutureSource(sourcePromise.future)
-          .watchTermination()(Keep.both)
-          .to(Sink.cancelled)
+          .toMat(Sink.fromSubscriber(probe))(Keep.left)
           .run()
 
       // wait for cancellation to occur
-      termination.futureValue should ===(Done)
+      probe.ensureSubscription()
+      probe.request(1)
+      probe.cancel()
 
       // even though canceled the underlying matval should arrive
       sourcePromise.success(underlying)
-      sourceMatVal.futureValue should ===("foo")
+      sourceMatVal.failed.futureValue.getMessage should ===("Stream finished before future source completed")
     }
 
     "fail if the underlying Future is failed" in assertAllStagesStopped {

--- a/akka-stream/src/main/scala/akka/stream/StreamDetachedException.scala
+++ b/akka-stream/src/main/scala/akka/stream/StreamDetachedException.scala
@@ -14,5 +14,4 @@ final class StreamDetachedException(message: String)
   with NoStackTrace {
 
   def this() = this("Stream is terminated. Materialized value is detached.")
-
 }

--- a/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
+++ b/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
@@ -22,6 +22,7 @@ import scala.annotation.unchecked.uncheckedVariance
 import scala.util.{ Failure, Success, Try }
 import scala.concurrent.duration.FiniteDuration
 import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
 
 /**
  * INTERNAL API
@@ -319,15 +320,10 @@ import scala.concurrent.{ Future, Promise }
 
           override def onDownstreamFinish(): Unit = {
             if (!materialized.isCompleted) {
-              // make sure we always yield the matval if possible, even if downstream cancelled
-              // before the source was materialized
-              val matValFuture = futureSource.map { gr ⇒
-                // downstream finish means it cancelled, so we push that signal through into the future materialized source
-                Source.fromGraph(gr).to(Sink.cancelled)
-                  .withAttributes(attr)
-                  .run()(subFusingMaterializer)
-              }(ExecutionContexts.sameThreadExecutionContext)
-              materialized.completeWith(matValFuture)
+              // we used to try to materialize the "inner" source here just to get
+              // the materialized value, but that is not safe and may cause the graph shell
+              // to leak/stay alive after the stage completes
+              materialized.tryFailure(new RuntimeException("Stream finished before future source completed"))
             }
 
             super.onDownstreamFinish()

--- a/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
+++ b/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
@@ -324,7 +324,7 @@ import scala.util.control.NonFatal
               // the materialized value, but that is not safe and may cause the graph shell
               // to leak/stay alive after the stage completes
 
-              materialized.tryFailure(new StreamDetachedException("Stream finished before future source completed"))
+              materialized.tryFailure(new StreamDetachedException("Stream cancelled before Source Future completed"))
             }
 
             super.onDownstreamFinish()

--- a/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
+++ b/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
@@ -323,7 +323,8 @@ import scala.util.control.NonFatal
               // we used to try to materialize the "inner" source here just to get
               // the materialized value, but that is not safe and may cause the graph shell
               // to leak/stay alive after the stage completes
-              materialized.tryFailure(new RuntimeException("Stream finished before future source completed"))
+
+              materialized.tryFailure(new StreamDetachedException("Stream finished before future source completed"))
             }
 
             super.onDownstreamFinish()

--- a/akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
+++ b/akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
@@ -182,13 +182,16 @@ object Source {
 
   /**
    * Streams the elements of the given future source once it successfully completes.
-   * If the future fails the stream is failed.
+   * If the [[Future]] fails the stream is failed with the exception from the future. If downstream cancels before the
+   * stream completes the materialized [[Future]] will be failed with a [[StreamDetachedException]].
    */
   def fromFutureSource[T, M](future: Future[_ <: Graph[SourceShape[T], M]]): javadsl.Source[T, Future[M]] = new Source(scaladsl.Source.fromFutureSource(future))
 
   /**
-   * Streams the elements of an asynchronous source once its given `completion` stage completes.
-   * If the `completion` fails the stream is failed with that exception.
+   * Streams the elements of an asynchronous source once its given [[CompletionStage]] completes.
+   * If the [[CompletionStage]] fails the stream is failed with the exception from the future.
+   * If downstream cancels before the stream completes the materialized [[CompletionStage]] will be failed
+   * with a [[StreamDetachedException]]
    */
   def fromSourceCompletionStage[T, M](completion: CompletionStage[_ <: Graph[SourceShape[T], M]]): javadsl.Source[T, CompletionStage[M]] =
     new Source(scaladsl.Source.fromSourceCompletionStage(completion))

--- a/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala
+++ b/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala
@@ -276,13 +276,16 @@ object Source {
 
   /**
    * Streams the elements of the given future source once it successfully completes.
-   * If the future fails the stream is failed.
+   * If the [[Future]] fails the stream is failed with the exception from the future. If downstream cancels before the
+   * stream completes the materialized `Future` will be failed with a [[StreamDetachedException]]
    */
   def fromFutureSource[T, M](future: Future[Graph[SourceShape[T], M]]): Source[T, Future[M]] = fromGraph(new FutureFlattenSource(future))
 
   /**
    * Streams the elements of an asynchronous source once its given `completion` stage completes.
-   * If the `completion` fails the stream is failed with that exception.
+   * If the [[CompletionStage]] fails the stream is failed with the exception from the future.
+   * If downstream cancels before the stream completes the materialized `Future` will be failed
+   * with a [[StreamDetachedException]]
    */
   def fromSourceCompletionStage[T, M](completion: CompletionStage[_ <: Graph[SourceShape[T], M]]): Source[T, CompletionStage[M]] = fromFutureSource(completion.toScala).mapMaterializedValue(_.toJava)
 

